File: leetcode/0318.go
package leetcode

func maxProduct(words []string) int {
	hash, ans := make(map[int]int), 0
	for _, word := range words {
		mask, length := 0, len(word)
		for _, ch := range word {
			mask |= 1 << (ch - 'a')
		}
		if length > hash[mask] {
			hash[mask] = length
		}
		for h_mask, h_len := range hash {
			if h_mask&mask == 0 {
				if ans < h_len*length {
					ans = h_len * length
				}
			}
		}
	}
	return ans
}

File: leetcode/0318.cpp
class Solution {
public:
    int maxProduct(vector<string>& words) {
        unordered_map<int, int> m;
        int ans = 0;
        for (auto& word : words) {
            int mask = 0, length = word.size();
            for (auto& ch : word) {
                mask |= 1 << (ch - 'a');
            }
            m[mask] = max(m[mask], length);
            for (auto& [h_mask, h_len] : m) {
                if (!(h_mask & mask)) {
                    ans = max(ans, length * h_len);
                }
            }
        }
        return ans;
    }
};